cmd/wallet-js-sdk/src/oidc/issuance/openid4ci.js

```diff
@@ -1,212 +1,287 @@
-  } else if (pre_authorized_code !== '') {
-    tokenResponse = await exchangePreAuthCode(pre_authorized_code, user_pin, transaction_data.issuer_metadata.token_endpoint);
-    // TODO: poll for token if authorization is pending
+  if (transaction_data.oauth_state !== oauth_state) {
+    throw new Error('Issuer provided a state parameter that does not match the state parameter of the client-side transaction. ' +
+      'This may be an error in the Issuer, or a mix-up of state data in a client performing multiple flows at the same time.')
   }
 
-  // stub
-  const sendCredentialRequest = async function(req) {
-    // data returned by issuer server
-    return {
-      format: '',
-      credential: {},
-    };
+  const tokenRequest = new URLSearchParams();
+  tokenRequest.append('grant_type', 'authorization_code');
+  tokenRequest.append('code', authorization_code);
+  tokenRequest.append('redirect_uri', client_config.wallet_callback_uri);
+  tokenRequest.append('client_id', client_config.client_id);
+
+  /*
+  response format:
+  {
+    access_token: '',
+    token_type: '',
+    c_nonce: '',
+  }
+  */
+  const token_response = await axios.post(
+    transaction_data.issuer_metadata.token_endpoint,
+    tokenRequest,
+    ).then((resp) => resp.data).catch((e) => {
+    throw new Error('Error authorizing wallet with Issuer: failed to exchange auth code for token:', e);
+  });
+
+  return await get_credential(token_response, transaction_data, client_config, jwt_signer);
+}
+
+async function preauthorized(
+  issuer_uri,
+  credential_type,
+  pre_authorized_code,
+  user_pin,
+  client_config,
+  jwt_signer = async function(iss, aud, iat, c_nonce){return ''},
+) {
+  // no op_state for pre-auth flow
+  // no oauth state for pre-auth flow
+
+  const issuer_metadata = await get_issuer_metadata(issuer_uri);
+  let transaction_data = {
+    issuer_metadata,
+    credential_type,
+    client_id: client_config.client_id,
   }
 
-  const jwt = await jwt_signer(transaction_data.client_id, transaction_data.issuer_metadata.issuer, /*time now*/0, tokenResponse.c_nonce)
+  const tokenRequest = new URLSearchParams();
+  tokenRequest.append('grant_type', 'urn:ietf:params:oauth:grant-type:pre-authorized_code');
+  tokenRequest.append('pre-authorized_code', pre_authorized_code);
+  tokenRequest.append('user_pin', user_pin);
+
+  /*
+  * tokenResponse contents look like:
+  * {
+      access_token: 'foobar',
+      token_type: 'bearer',
+      c_nonce: 'sldkfyhsiudlfkgjh',
+      authorization_pending: true,
+      interval: 5,
+    }
+  * */
+  // TODO: POST as application/x-www-form-urlencoded not json
+  let token_response = await axios.post(
+    transaction_data.issuer_metadata.token_endpoint,
+    tokenRequest,
+  ).then((resp)=> resp.data);
+  // TODO: poll if response is deferred (if token_response.authorization_pending is true),
+  //  waiting token_response.interval seconds before next request (or 5 seconds if interval missing or less than 5).
+
```

Comment:
We should give the control to consumer to poll as deferred flow could take days as well.